src/webgpu/shader/execution/robust_access_vertex.spec.ts

```diff
@@ -207,181 +239,203 @@ class DrawCall {
+    type: GPUVertexFormat
+  ) {
+    const typeInfo = typeInfoMap[type];
     // Vertex buffer descriptors
     const buffers: GPUVertexBufferLayout[] = [];
     {
       let currAttribute = 0;
-      for (let i = 0; i < bufferContents.length; i++) {
+      for (let i = 0; i < bufferCount; i++) {
         buffers.push({
-          arrayStride: attributesPerBuffer * typeInfo.size,
+          arrayStride: attributesPerBuffer * typeInfo.sizeInBytes,
           stepMode: i === 0 ? 'instance' : 'vertex',
           attributes: Array(attributesPerBuffer)
             .fill(0)
             .map((_, i) => ({
               shaderLocation: currAttribute++,
-              offset: i * typeInfo.size,
-              format: p.type as GPUVertexFormat,
+              offset: i * typeInfo.sizeInBytes,
+              format: type as GPUVertexFormat,
             })),
         });
       }
     }
+    return buffers;
+  }
 
-    // Offset the range checks for gl_VertexIndex in the shader if we use BaseVertex
-    let vertexIndexOffset = 0;
-    if (p.drawCallTestParameter === 'baseVertex') {
-      vertexIndexOffset += p.errorScale;
+  generateVertexShaderCode({
+    bufferCount,
+    attributesPerBuffer,
+    validValues,
+    typeInfo,
+    vertexIndexOffset,
+    numVertices,
+    isIndexed,
+  }: {
+    bufferCount: number;
+    attributesPerBuffer: number;
+    validValues: number[];
+    typeInfo: VertexInfo;
+    vertexIndexOffset: number;
+    numVertices: number;
+    isIndexed: boolean;
+  }): string {
+    // Create layout and attributes listing
+    let layoutStr = 'struct Attributes {';
+    const attributeNames = [];
+    {
+      let currAttribute = 0;
+      for (let i = 0; i < bufferCount; i++) {
+        for (let j = 0; j < attributesPerBuffer; j++) {
+          layoutStr += `[[location(${currAttribute})]] a_${currAttribute} : ${typeInfo.wgslType};\n`;
+          attributeNames.push(`a_${currAttribute}`);
+          currAttribute++;
+        }
+      }
     }
+    layoutStr += '};';
+
+    const vertexShaderCode: string = `
+      ${layoutStr}
+
+      fn valid(f : f32) -> bool {
+        return ${validValues.map(v => `f == ${v}.0`).join(' || ')};
+      }
+
+      fn validationFunc(v : ${typeInfo.wgslType}) -> bool {
+        ${typeInfo.validationFunc}
+      }
+
+      [[stage(vertex)]] fn main(
+        [[builtin(vertex_index)]] VertexIndex : u32,
+        attributes : Attributes
+        ) -> [[builtin(position)]] vec4<f32> {
+        var attributesInBounds : bool = ${attributeNames
+          .map(a => `validationFunc(attributes.${a})`)
+          .join(' && ')};
+
+        var indexInBoundsCountFromBaseVertex : bool =
+            (VertexIndex >= ${vertexIndexOffset}u &&
+            VertexIndex < ${vertexIndexOffset + numVertices}u);
+        // For D3D12, VertexIndex count from 0. For index robust access it can be 0.
+        var indexInBoundsCountFrom0 : bool = (VertexIndex >= 0u && VertexIndex < ${numVertices}u);
```

Comment:
nit: here and in other places we don't need type annotations.

Reply:
Done.